src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -1433,6 +1432,34 @@ bool is_superset(
     return true;
 }
 
+// this function assumes that the inputs are valid schemas, and if they have a
+// $schema member it will have a string value
+bool check_compatible_dialects(
+  json::Value const& older, json::Value const& newer) {
+    auto get_dialect =
+      [](json::Value const& v) -> std::optional<json_schema_dialect> {
+        if (!v.IsObject()) {
+            // support true/false schemas
+            return std::nullopt;
+        }
+        auto it = v.FindMember("$schema");
+        if (it == v.MemberEnd()) {
+            return std::nullopt;
+        }
+        return from_uri(it->value.GetString());
+    };
+
+    auto older_dialect = get_dialect(older);
+    auto newer_dialect = get_dialect(newer);
+    if (!older_dialect.has_value() || !newer_dialect.has_value()) {
```

Comment:
given this test case:
```cpp
  {
    .reader_schema = R"({
    "$schema": "http://json-schema.org/draft-04/schema",
    "type": "integer",
    "minimum": 10,
    "exclusiveMinimum": true
    })",
    .writer_schema = R"({
    "type": "integer",
    "minimum": 10,
    "exclusiveMinimum": 10
    })",
    .reader_is_compatible_with_writer = false,
    .expected_exception = true,
  },
```
one is explicitly draft4, the other is implicitly the latest draft

My reasoning was that this should be safe, because even though check_compatible_dialect returns true, is_superset would block this because we don't handle `exclusiveMinimum` changing type (or some other check, like not implemented keywords).

but i changed back to a stricter &&